designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+* `plugins` - replaced by `ruledefs`
+* `root` - always considered `true`
+
+Each of these keys represent different ways of augmenting how configuration is calculated and all of that responsibility now falls on the user.
+
+#### Referencing Plugin Rules
+
+The `plugins` key in `.eslintrc` was an array of strings indicating the plugins to load, allowing you to specify processors, rules, etc., by referencing the name of the plugin. It's no longer necessary to indicate the plugins to load because that is done directly in the `.eslint.js` file. For example, consider this `.eslintrc` file:
+
+```yaml
+plugins:
+  - react
+
+rules:
+  react/jsx-uses-react: error
+```
+
+This file tells ESLint to load `eslint-plugin-react` and then configure a rule from that plugin. The `react/` is automatically preprended to the rule by ESLint for easy reference.
+
+In `.eslint.js`, the same configuration is achieved using a `ruledefs` key:
+
+```js
+const reactPlugin = require("eslint-plugin-react");
+
+exports.config = {
+    ruledefs: {
+        react: reactPlugin.rules
+    },
+    rules: {
+        "react/jsx-uses-react": "error"
+    }
+};
+```
+
+Here, it is the `ruledefs` that assigns the name `react` to the rules from `eslint-plugin-react`. The reference to `react/` in a rule will always look up that value in the `ruledefs` key.
+
+#### Referencing Parsers and Processors
+
+In `.eslintrc`, the `parser` and `processor` keys required strings to be specified, such as:
+
+```yaml
+plugins: ["markdown"]
+parser: "babel-eslint"
+processor: "markdown/markdown"
+```
+
+In `.eslint.js`, you would need to pass the references directly, such as:
+
+```js
+exports.config = {
+    parser: require("babel-eslint"),
+    processor: require("eslint-plugin-markdown).processors.markdown
+};
+```
+
+In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never of question of where the modules will be resolved from.
+
+#### Applying an Environment
+
+Unlike with `.eslintrc` files, there is no `env` key in `.eslint.js`. Users can mimmick the behavior of `env` by assigning directly to the `globals` key:
+
+```js
+const globals = require("globals");
+
+exports.config = {
+    globals: {
+        MyGlobal: true,
+        ...globals.browser
+    }
+};
+```
+
+This effectively duplicates the use of `env: { browser: true }` in ESLint.
+
+**Note:** This would allow us to stop environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.
+
+#### Extending Another Config
+
+With the removal of `extends`, it's now up to the users to inherit from other configs. That can easily be done in the following manner:
+
+```js
+const standardConfig = require("eslint-config-standard");
+
+standardConfig.rules.semi = ["error", "always"];
+
+exports.config = standardConfig;
```

Comment:
Would the user still need to add a `files` key to the shareable config? Presumably the files being linted are project-specific.

Reply:
Good point! Shareable configs shouldn't be providing `files` or `ignores`. I'll add mention of that.

Reply:
Actually, upon reading this section over again, I can see a use case for shareable configs providing `files` and `ignore`. Shareable configs can be general, like the Airbnb style guide, or company-specific, in which case they might want to have these fields standardized.

Reply:
Not sure if the extra complexity is worth it, but-- could/should we add either a config key, a CLI/CLIEngine option, or an argument to `Config.merge` etc., that will opt-in or opt-out of receiving globs from extended configs during config merge?

Reply:
My thinking was that people who didn't want to import `files` and `ignore` would just specify their own patterns that would overwrite the ones they were extending. We could certainly add a utility to help pick things out, but I think the necessity is based on how we decide to merge those fields. If we always overwrite, then I don't think we need a utility for this purpose.

Also, see @mysticatea's [comment below](https://github.com/eslint/rfcs/pull/9#discussion_r250037797) about how he uses file patterns in shareable configs currently.